### ISSUE TITLE
Panic in len_zero

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/rust-clippy#2019]
Clippying the current master branch of rayon, https://github.com/nikomatsakis/rayon/commit/ea50fb49c27d1019345d225699a263455f5f5ae1.

```
$ RUST_BACKTRACE=1 cargo +nightly-2017-09-02 clippy
   Compiling rayon v0.8.2
thread 'rustc' panicked at 'the DefId is local, the NodeId should be available', /checkout/src/libcore/option.rs:839:4
stack backtrace:
  10: <core::option::Option<T>>::expect
             at /checkout/src/libcore/option.rs:302
  11: clippy_lints::len_zero::check_trait_items::fill_trait_set
             at .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.156/src/len_zero.rs:112
  12: clippy_lints::len_zero::check_trait_items::fill_trait_set
             at .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.156/src/len_zero.rs:117
  13: clippy_lints::len_zero::check_trait_items
             at .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.156/src/len_zero.rs:130
  14: <clippy_lints::len_zero::LenZero as rustc::lint::LateLintPass<'a, 'tcx>>::check_item
             at .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.156/src/len_zero.rs:71
error: Could not compile `rayon`.
```

**Comment by dtolnay** [comment]
Minimal repro:

```rust
pub trait ParallelIterator: Sized {}

pub trait IndexedParallelIterator: ParallelIterator {
    fn len(&mut self) -> usize;
}
```

**Comment by Manishearth** [comment]
It's the panic at https://github.com/rust-lang-nursery/rust-clippy/blob/master/clippy_lints/src/len_zero.rs#L115

That expect seems correct though. Hm.

**Comment by oli-obk** [comment]
The issue is the `Sized` bound. Or any bound on a trait from outside the current crate. We should probably be storing the `DefId` and using [`TyCtxt::provided_trait_methods`](http://manishearth.github.io/rust-internals-docs/rustc/ty/struct.TyCtxt.html#method.provided_trait_methods) to iterate through the methods instead of going through the HIR.

**Comment by oli-obk** [comment]
fixed in 0.0.157

**Comment by dtolnay** [comment]
Thanks! Just linking the fix: https://github.com/rust-lang-nursery/rust-clippy/commit/7757c893ef89690e52ff71dd7ee65d02edd8ea49.
